facilitator/src/gcp_oauth.rs

```diff
@@ -84,17 +91,231 @@ struct ServiceAccountKeyFile {
     token_uri: String,
 }
 
+/// This enum captures the different ways that a GcpOauthTokenProvider can
+/// obtain the default Oauth token, used either to authenticate to GCP services
+/// or to obtain a further service account Oauth token from GCP IAM.
+enum DefaultTokenProvider<P: ProvideAwsCredentials> {
+    /// Fetch a token from the GKE metadata service for the GCP SA mapped to
+    /// the current Kubernetes SA via GKE workload identity (not to be confused)
+    /// with workload identity *pool*). Used by workloads in Google Kubernetes
+    /// Engine.
+    GkeMetadataService,
+    /// Use a GCP service account key file to authenticate to GCP IAM as some
+    /// service account.
+    ServiceAccountKeyFile(ServiceAccountKeyFile),
+    /// Use a GCP Workload Identity Pool to federate GCP IAM with AWS IAM,
+    /// allowing workloads in AWS Elastic Kubernetes Service to impersonate a
+    /// GCP SA.
+    AwsIamFederationViaWorkloadIdentityPool(WorkloadIdentityPoolParameters<P>),
+}
```

Comment:
In another project I had an enum like this, but I would up being much happier making it a trait. Have you considered making this a trait instead? So then instead of `default_token` providing a match-based dispatch to `account_token_from_gke_metadata_service`, `account_token_with_key_file`, etc, you'd have a trait with one required method, `default_token`, that would be implemented by each of `GkeMetadataService` (a new unit type), `ServiceAccountKeyFile`, and `WorkloadIdentityPoolParameters<P>`.

Also I note that the dispatched-to methods don't actually seem to use `&self`; the trait approach would make more meaningful use of `&self`.

Reply:
I think I did this because I was replicating what I did in `aws_credentials::Provider`, where I provided an explanation for this pattern: https://github.com/abetterinternet/prio-server/blob/c0813359f01937d8861bdff70a6c3caea2ad379d/facilitator/src/aws_credentials.rs#L39

Those constraints don't really apply in this case. I'll give it a try with a trait object member on `GcpOauthTokenProvider` and see if I like it better.

Reply:
Ultimately I like it better your way. Besides the advantages you describe, I also used `Box<dyn ProvideAwsCredentials>`, which means that several structs no longer need to be generic over `P: ProvideAwsCredentials`, at the cost of some dynamic dispatch. I think that's better than introducing AWS-related type parameters to numerous structs that have nothing (obvious) to do with AWS, though.